src/Api.js

```diff
@@ -89,13 +92,68 @@ class Api {
 	}
 
 	/**
-	 * Get the WhoColor data for a given wiki page.
+	 * Get the WikiWho data for a given wiki page. Use onError and onResults to set return handlers.
 	 *
-	 * @param  {string} wikiUrl URL of the wiki page that we want to analyze.
-	 * @return {Promise}
+	 * @param {string} wikiUrl URL of the wiki page that we want to analyze.
 	 */
 	getData( wikiUrl ) {
-		return $.getJSON( this.getAjaxURL( wikiUrl ) );
+		$.getJSON( this.getAjaxURL( wikiUrl ) ).then( ( function ( result ) {
```

Comment:
if you return this promise, you won't need to do the trick of attaching `.onError`

So, if the process here cleans up the results, you could return the promise here, and then the process on `app.js` can be clean. 

Something like this in API:

```
return $.getJSON( ... )
    .then(
        // success
       function () {
             // On error, return a rejected promise with the error code/name/whatever we choose
             return $.Deferred().reject( errorCode );
             // On success, return the new cleaned up data
             return cleanedResults;
       },
       // API error
       function () {
            // Any API error like timing out, etc, should pop up in the failure state, so it can be processed to give
            // us some errorCode
            return errorCode
       }
     );
```

Then, in `app.js` you can "just" do this:

```
components.api.getData( ... )
    .then(
         function ( goodData ) {
               // Toggle the DOM, replace the HTML, etc etc
         },
         function ( errorCode ) {
              // Alll error codes, including the ones that started with supposed success
             // are now rejected promise, which makes it easier for us to now treat this as
             // a regular/common standardized error and display a message key in the InfoBarWidget
       }
);
```

This can make this much cleaner **and** separate concerns a lot more; it would mean that the API class deals with whatever funky ways wikiwho displays errors (even if the promise succeeded, we force-reject it, so the output from the API class gives us standard and cleaner way to handle success vs fails)

Reply:
You can use a member variable, especially since it looks like the getData method is more about "wait until I'm done" rather than "let me give you the data you need".

In general, though, in order to deliver the results onwards, you "just" need to return them from within the promise. So, in this case, you could just `return goodData;` at the end. This would mean that whatever attaches next to this promise, has that parameter for the success --

```
api.getData( ... ).then( result => { /* result is the "goodData" you returned */ } );
```

This is more for general future/knowledge/discussion. I think the strategy you go with is also fine, and means there's no need to return the values. I wrote above that I have a preference towards doing everything in the promise (returning the values and using them as your starting point) but that, too, has some edge case price, so it's not really crucial which way we choose. I think the way you went with is good.